### ISSUE TITLE
Crash on song change or while disconnecting - OpenBSD platform

### ISSUE THREAD

**Issue by arcadilius** [mikebrady/shairport-sync#975]
I've followed setup instructions and was able to use it for a brief moment, great work by the way! However, after song change or after changing output back to laptop speakers shairport crashes. After tweaking this and that, with very small buffer on sndiod I've got it a bit more stable but annoying issue persists. My source is Macbook Pro or iPhone. Both tested machines are latest patched OpenBSD's with Intel HDA. I can push 50MB/s over wi-fi. I understand that OpenBSD is not Tier 1 platform yet but maybe there's something I'm missing?

**Comment by mikebrady** [comment]
Thanks for the report. I’ll take a look.

**Comment by mikebrady** [comment]
So I’ve been testing using a non-GUI installation of OpenBSD in a VM. Could you say what kind of installation you are using, please? And if you have any pointers to installation and configuration, it would be appreciated.

**Comment by arcadilius** [comment]
Both machines are headless file server and router, both have 8GB of RAM and Intel HDA's, two and four physical Intel cores with more than enough processing power, n3150 and I5-560M. Environment is not crowdy as it's running in home network on mesh with 1300Mbit/s - transfers over wifi at 700Mbit/s with average 4ms response time - distance between mesh points up to 5 meters in direct visibility - not a packet dropped. No crazy system tweaks except load optimisations targeted for snappy file transfers and bufferbloat mitigations on router exclusively. Sound servers in both configuration were reporting 0 packets lost and shairport process was in up to 5% cpu time range. Tested in several scenarios, root, privileged user and user with limited access (this last one created for running service at boot). Like mentioned earlier - nothing that could affect network performance varies from default OpenBSD installation. Tested software with soxr flag as well - same effect. For comparison, I was running shairport on Raspberry Pi first generation before without hiccups. At this point I'm running out of ideas, however I'm planning to set up a Linux guest in VMM and push sound from ALSA over the network to SNDIO server - that's how bad I want this running in my bedroom :)

**Comment by mikebrady** [comment]
Thanks for all that. I've found what looks like a good guide to installing a full OpenBSD and will see what I can find. The last version I used was 6.2.

**Comment by mikebrady** [comment]
So, I can get it to crash alright. It seems to be getting some new signals back from the `sndio` subsystem. I'll take a look at it, but it might take a little while.

**Comment by arcadilius** [comment]
Take your time. I do appreciate you help.

**Comment by mikebrady** [comment]
There may something amiss in the `sndio` subsystem in OpenBSD. It's working very well in FreeBSD. Of course, it may be that `sndio` is working properly and Shairport Sync is misusing it somehow. I have emailed the `sndio` mailing list for help.

**Comment by arcadilius** [comment]
That sounds like OpenBSD devs typical approach. It's a common knowledge that they tend to make changes between releases that brake things and render porting a pain. On a side note, I've used your software on an oldest SBC I've had, first generation RBPi - smooth like butter. Thanks again for trying to solve the mystery.

**Comment by mikebrady** [comment]
Okay, so the plot thickens. I could get Shairport Sync to crash on OpenBSD 6.6 running in a virtual machine (VMWare Fusion on an iMac). However, on the suggestion of the `sndio` developer, I moved to a real machine. It is an old Lenovo G560 which has "Intel 3400 HD Audio". In short, everything works perfectly. Excellent synchronisation too.

It is well known that the emulations of sound cards in VMs are problematic – I know this myself and I should have thought of it, though FreeBSD 12.1 works in the same environment.

All in all, it seems to point to a problem with the support of the Intel HDAs on your system, I suppose.

**Comment by arcadilius** [comment]
I've never experienced sound problems before as both my machines are running without graphical interfaces - never made a sound, maybe except beep at boot. Even with both cards detected and working on first sight there's a lot to wish for as it turned out. Without your suggestions I would never start digging deeper to find out that this is in fact driver issue and lack of full support on kernel side - I'm convinced now that sndio is not to blame, nor it's shaiport's fault. On a side note - FreeBSD got this problem sorted but I'm far from switching to another OS. I don't think I have strength, or knowledge, to amend kernel code to make it work and recompile every time there's patch available. There are some suggestions what could be done in azalia code but to be honest - playing music is not the sole purpose of these machines, would be just nice addition. I do apologise I've wasted your time.

**Comment by mikebrady** [comment]
Thanks for your response. Given that `sndio` is native to OpenBSD, they might be interested in your experience — it might be an opportunity for them to troubleshoot an OpenBSD-related issue. Meantime, perhaps we should close this issue for the moment.

**Comment by mikebrady** [comment]
One other slightly ironic thing is that `sndio` is available for Linux and runs very well on the Raspberry Pi.

**Comment by arcadilius** [comment]
To summarise, I've tried everything... I'm not a quitter but I'm calling it in this case. I had similar if not the same issue on two different boards and I could not believe this may be OS related. OpenBSD is amazing platform but in my case turned out to be too hardware sensitive. Long story short - after installing FreeBSD on one of machines I've noticed not only that sound was working perfectly out of the box (not to mention sound equaliser was built in - just needed a device hint) but overall performance was noticeably better. Thank you for shown interest and support. Love you work, keep it up. It's time for a closure.
